sdk/client-core/core/src/main/java/io/clientcore/core/http/client/DefaultHttpClient.java

```diff
@@ -418,18 +419,57 @@ private static void buildAndSend(HttpRequest httpRequest, OutputStreamWriter out
 
         /**
          * Reads the response from the input stream and extracts the information needed to construct an instance of
-         * HttpUrlConnectionResponse
+         * Response
          *
          * @param httpRequest The HTTP Request being sent
          * @param reader the input stream from the socket
-         * @return an instance of HttpUrlConnectionResponse
+         * @return an instance of Response
          * @throws IOException If an I/O error occurs
          */
         private static Response<?> buildResponse(HttpRequest httpRequest, BufferedReader reader) throws IOException {
+            // Parse Http response from socket:
+            // Status Line
+            // Response Headers
+            // Blank Line
+            // Response Body
+
+            int statusCode = readStatusCode(reader);
+
+            HttpHeaders headers = readResponseHeaders(reader);
+            HttpHeader contentLengthHeader = headers.get(CONTENT_LENGTH);
+            String line;
+            // read body if present
+            if (headers.getSize() != 0 && contentLengthHeader != null && contentLengthHeader.getValue() != null) {
+                StringBuilder bodyString = new StringBuilder();
+                while ((line = reader.readLine()) != null) {
+                    bodyString.append(line);
+                }
+
+                return new HttpResponse<>(httpRequest, statusCode, headers,
+                    BinaryData.fromString(bodyString.toString()));
+            }
+            return new HttpResponse<>(httpRequest, statusCode, headers, null);
+        }
+
+        private static int readStatusCode(BufferedReader reader) throws IOException {
+
+            while (!reader.ready()) {
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
             String statusLine = reader.readLine();
+
+            if (statusLine == null) {
+                throw LOGGER.logThrowableAsError(new IllegalStateException("Unexpected response from server."));
+            }
             int dotIndex = statusLine.indexOf('.');
-            int statusCode = Integer.parseInt(statusLine.substring(dotIndex + 3, dotIndex + 6));
+            return Integer.parseInt(statusLine.substring(dotIndex + 3, dotIndex + 6));
```

Comment:
We should add further validation here:

- Need to make sure we're using HTTP/1.0 or 1.1
- While slower, I wouldn't use a magic offset from the '.' index given the status line is a well specific format: https://www.rfc-editor.org/rfc/rfc2616#section-6.1